### ISSUE TITLE
serve sftp: serving a backend which doesn't support MD5 hashes gives errors

### ISSUE THREAD

**Issue by Forum1877621** [rclone/rclone#6351]
#### What is the problem you are having with rclone?

I encountered an issue when file names have periods (under specific circumstances)

The error:

```
2022/07/31 14:42:30 ERROR : drive/tmp/test ... test ... test.mp4: WriteFileHandle: Truncate: Can't change size without --vfs-cache-mode >= writes
2022/07/31 14:42:31 ERROR : drive/tmp/test ... test ... test.mp4: WriteFileHandle: ReadAt: Can't read and write to file without --vfs-cache-mode >= minimal
2022/07/31 14:42:32 ERROR : drive/tmp/test ... test ... test.mp4: Failed to calculate dst hash: failed to calculate md5 hash: failed to run "md5sum /drive/tmp/test\\ ...\\ test\\ ...\\ test.mp4": hash failed: hash type not supported: Process exited with status 1
2022/07/31 14:42:32 ERROR : drive/tmp/test ... test ... test.mp4: corrupted on transfer
2022/07/31 14:42:32 ERROR : drive/tmp/test ... test ... test.mp4: WriteFileHandle.New Rcat failed: corrupted on transfer
2022/07/31 14:42:32 ERROR : drive/tmp/test ... test ... test.mp4: WriteFileHandle.Flush error: corrupted on transfer
2022/07/31 14:42:32 ERROR : IO error: corrupted on transfer
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
2022/07/31 14:42:37 ERROR : IO error: can't open file - writer failed
```

How to reproduce:

- download any video and put some periods in it its name (for example, "test ... test ... test.mp4"
- on one Windows computer, serve a crypt with any filename encoding over SFTP
- on another Windows computer mount the crypt
- copy the file to the crypt on the computer that mounted the crypt
- errors will appears
- try to play the video
- errors will appear, the video won't play

This error doesn't seem to occur if the crypt is a local folder, I've only seen it happen when copying files with periods in their names over SFTP into a mounted crypt.

#### What is your rclone version (output from `rclone version`)

rclone v1.59.0
- os/version: Microsoft Windows 11 21H2 (64 bit)
- os/kernel: 10.0.22000.795 (x86_64)
- os/type: windows
- os/arch: amd64
- go/version: go1.18.3
- go/linking: static
- go/tags: cmount

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Windows 10, 64 bit

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Is the problem with the shell escaping to call the md5sum command?

    drive/tmp/test ... test ... test.mp4: Failed to calculate dst hash: failed to calculate md5 hash: failed to run "md5sum /drive/tmp/test\\ ...\\ test\\ ...\\ test.mp4": hash failed: hash type not supported: Process exited with status 1

`rclone serve` has its own internal implementation of md5sum so maybe it is decoding the shell escaping incorrectly on Windows.

@albertony what do you think?

@Forum1877621 it would be very helpful if you could show the exact rclone commands you are using and the rclone config with secrets redacted - thanks.

**Comment by albertony** [comment]
@ncw: Yeah, that path looks strange: `test\\ ...\\ test\\ ...\\ test.mp4`. I think the double `\\` is simply the log message formatting it with `%q` (and not `%s` or `%v`, which I think would be better in many cases, and instead surround with quotes manually). Then what remains is backslash escaped space characters. That is not very Windows-y... And sure enough, if I test with `--sftp-shell-type=unix` I get similar problem. So rclone is treating the server as having "Unix shell". 

@Forum1877621: How is the config of the sftp remote in rclone, especially the `shell_type` option? It will be auto-detected on first use of the remote and saved back to config file, but if the detection fails it will default to `unix` and you will have to manually fix this. You can test overriding it temporarily by adding `--sftp-shell-type=cmd` (or `--sftp-shell-type=powershell` if your sftp server is configured with PowerShell instead) to your rclone command. If this is the case, i.e. the auto-detection failed, it would be interesting to see a debug log from it: Remove any `shell_type` option from your remote config, or create a brand new remote for testing, and run a simple command with verbose logging, e.g. `rclone about yoursftpremote: -vv`

**Comment by Forum1877621** [comment]
**Important note: I've only noticed this problem when serving a crypt (with any file encoding)**

@albertony

I think the autodetection might be broken on windows. On both of my Windows computers the shell_type gets set to unix.

```
[test]
type = sftp
host = 192.168.2.50
user = my-user
port = 54321
pass = some-password
md5sum_command = md5sum
sha1sum_command = sha1sum
shell_type = unix
```

rclone about doesn't seem to work

```
...\rclone>rclone.exe about test -vv
2022/08/01 17:13:25 DEBUG : rclone: Version "v1.59.0" starting with parameters ["rclone.exe" "about" "test" "-vv"]
2022/08/01 17:13:25 DEBUG : Creating backend with remote "test"
2022/08/01 17:13:25 DEBUG : Using config file from "C:\\Users\\...\\AppData\\Roaming\\rclone\\rclone.conf"
2022/08/01 17:13:25 DEBUG : fs cache: renaming cache item "test" to be canonical "//?/D:/some/path/some/rclone/test"
2022/08/01 17:13:25 DEBUG : 2 go routines active
2022/08/01 17:13:25 Failed to about: about call failed: failed to read disk usage: The system cannot find the path specified.
```

If I delete the shell_type line and run a command like so:

```
...\rclone>rclone.exe -v --stats=5s mount test:/ "path\test" --cache-dir="E:\vfs" --vfs-cache-mode=writes --vfs-cache-max-size=10000M --vfs-cache-max-age=72h
2022/08/01 17:09:34 INFO  : sftp://my-user@192.168.2.51:54321//: poll-interval is not supported by this remote
2022/08/01 17:09:34 INFO  : vfs cache: cleaned: objects 0 (was 0) in use 0, to upload 0, uploading 0, total size 0 (was 0)
2022/08/01 17:09:34 ERROR : sftp://my-user@192.168.2.51:54321//: Statfs failed: your remote may not have the required df utility: failed to run "df -k /": about failed: failed to read disk usage: The system cannot find the path specified.: Process exited with status 1
The service rclone has been started.
```

the shell_type gets set to unix again, not cmd.

I tested copying file with periods when shell_type is cmd but I get the same error

@ncw 

config is above

Commands

Windows PC 1: make a crypt

```
[my-crypt]
type = crypt
remote = E:\some-folder
password = a
password2 = b
filename_encoding = base64
```

Windows PC 1: rclone.exe -vv --stats=5s serve sftp my-crypt:/ --user=user-1 --pass=password-1 --addr=:54321

Windows PC 2: rclone.exe -vv --stats=5s --file-perms=0777 mount the-crypt:/ "...\test-mount-point"

Then, on PC 2 copy "test ... test ... test.mp4" to the mounted folder and look at the console output. You should see the error.

**Comment by albertony** [comment]
@Forum1877621 Thanks for testing.

> rclone about doesn't seem to work
> ...\rclone>rclone.exe about test -vv

It seems you forgot a `:` following your remote name in your command in this test, it should be:

```
...\rclone>rclone.exe about test: -vv
```

> If I delete the shell_type line and run a command like so:
> ...\rclone>rclone.exe -v --stats=5s mount ...

Could you please show the debug log in this test, i.e. double up to `-vv`. If you remove the shell_type option and run the next (any) command with `-vv` it should print the output of the shell detection code as debug messages. That will be really interesting to see.

> I think the autodetection might be broken on windows.

In any case you can configure it yourself. It is under advanced config, so when you set up your remote with `rclone config` you will have to answer `y` on the question to set advanced options. The auto detection done by default was never bullet proof and meant more as a convenience feature that would work for most users. I would really like to see why it fails in your case, though.

**Comment by ncw** [comment]
@Forum1877621 what is the sftp server? is it `rclone serve sftp`?

**Comment by Forum1877621** [comment]
@ncw yes

rclone.exe -vv --stats=5s serve sftp my-crypt:/ --user=user-1 --pass=password-1 --addr=:54321

**Comment by albertony** [comment]
Aha, that is relevant information, I didn't catch that...

The `serve sftp` ("rclone server") is treated as a unix shell by the sftp backend ("rclone client"), regardless what it is running on (Windows or Linux), because it has code that detects typical unix commands such as `df` and `md5sum` and returns proper result for them without actually executing the corresponding commands. It also has code to "unescape" unix shell style escaping, i.e. backslash escapes. So `shell_type = unix` should be correct after all. But for some reason the escaping and/or unescaping trips up on something. Did a simple test locally, paths with spaces, periods etc, and that worked fine. I will run some more tests tomorrow, but in any case debug log output (`-vv`) from your original case would still be beneficial, if possible from both sides so that we can try to track how the escaping and unescaping is being done.

**Comment by Forum1877621** [comment]
@albertony 

I removed the shell_type again and ran the following command

`rclone.exe -vv about some-sftp:`

Here's the output

```
...rclone...\>rclone.exe -vv about some-sftp:
2022/08/02 21:43:53 DEBUG : rclone: Version "v1.59.0" starting with parameters ["rclone.exe" "-vv" "about" "some-sftp:"]
2022/08/02 21:43:53 DEBUG : Creating backend with remote "some-sftp:"
2022/08/02 21:43:53 DEBUG : Using config file from "----\\Roaming\\rclone\\rclone.conf"
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: New connection 192.168.2.49:8023->192.168.2.50:54321 to "SSH-2.0-rclone/v1.59.0"
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Running shell type detection remote command: echo ${ShellId}%ComSpec%
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Remote command result: ${ShellId}%ComSpec%
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Shell type "unix" detected (set option shell_type to override)
2022/08/02 21:43:53 DEBUG : Saving config "shell_type" in section "some-sftp" of the config file
2022/08/02 21:43:53 DEBUG : Keeping previous permissions for config file: -rw-rw-rw-
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Relative path resolved to "/"
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Using root directory "/"
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Server has VFS statistics extension
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: About path "/"
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Failed to retrieve VFS statistics, trying shell command instead: sftp: "operation unsupported" (SSH_FX_OP_UNSUPPORTED)
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Shell path "/"
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: About path "/"
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: About using shell command for shell type "unix"
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Running remote command: df -k /
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: About shell command for shell type "unix" failed (set option shell_type to override): failed to run "df -k /": about failed: failed to read disk usage: The system cannot find the path specified.: Process exited with status 1
2022/08/02 21:43:53 DEBUG : 12 go routines active
2022/08/02 21:43:53 DEBUG : sftp://user@192.168.2.50:54321/: Closing 1 unused connections
2022/08/02 21:43:53 Failed to about: about call failed: your remote may not have the required df utility: failed to run "df -k /": about failed: failed to read disk usage: The system cannot find the path specified.: Process exited with status 1
```

**Comment by Forum1877621** [comment]
> But for some reason the escaping and/or unescaping trips up on something. Did a simple test locally, paths with spaces, periods etc, and that worked fine. 

I haven't had any problems with local files and periods. I also haven't had problems serving a normal folder over SFTP. I've only seen the problem when serving a crypt. Try to serve a crypt over SFTP and on another computer copy a video with periods in the file name to the crypt. If you look at the log output you should see errors.

**Comment by albertony** [comment]
Right. I reproduced it now.

Don't think it is related to shell escaping.

When serving a crypt remote, the serve(r) returns "hash type not supported" from here:
https://github.com/rclone/rclone/blob/3a8e52de74e81f499f44a47a937552378d5bc83b/backend/crypt/crypt.go#L904

Called from here:
https://github.com/rclone/rclone/blob/3a8e52de74e81f499f44a47a937552378d5bc83b/cmd/serve/sftp/connection.go#L125

Will have to dig deeper into it later. Perhaps @ncw in the mean time sees💡 ... 😄

**Comment by albertony** [comment]
@ncw I don't understand these two:

https://github.com/rclone/rclone/blob/3a8e52de74e81f499f44a47a937552378d5bc83b/backend/crypt/crypt.go#L901-L905

https://github.com/rclone/rclone/blob/3a8e52de74e81f499f44a47a937552378d5bc83b/backend/crypt/crypt.go#L1042-L1063

Why is Object.Hash not implemented? Does not the Object type embed ObjectInfo, so if ObjectInfo can do Hash then Object should be able to do the same?

**Comment by albertony** [comment]
And also, why does this end up in Object.Hash instead of ObjectInfo.Hash:

https://github.com/rclone/rclone/blob/9d3958bd0b0f26ee597cb732761d705913bacdb1/cmd/serve/sftp/connection.go#L121-L125

**Comment by ncw** [comment]
I think the problem is that this code is no longer being activated as we've changed the md5sum detection code

https://github.com/rclone/rclone/blob/918bd6d3c3b28606e0208ee657b25274e4a06a18/cmd/serve/sftp/connection.go#L137-L154

If that was activated then the md5sum test would fail and it wouldn't attempt an MD5 checksum.

Something like this does fix the problem, perhaps in a more general way as it returns an error however md5sum is called.

```patch
--- a/cmd/serve/sftp/connection.go
+++ b/cmd/serve/sftp/connection.go
@@ -101,6 +101,10 @@ func (c *conn) execCommand(ctx context.Context, out io.Writer, command string) (
 		if binary == "sha1sum" {
 			ht = hash.SHA1
 		}
+		hashes := c.vfs.Fs().Hashes()
+		if !hashes.Contains(ht) {
+			return fmt.Errorf("%v hash not supported", ht)
+		}
 		var hashSum string
 		if args == "" {
 			// empty hash for no input
```

@albertony do we need to special case the md5sum/sha1sum detection in rclone serve sftp?

**Comment by albertony** [comment]
> I think the problem is that this code is no longer being activated as we've changed the md5sum detection code

Aha... Found that change here: https://github.com/rclone/rclone/commit/3ea82032e796d1f107f7c049cbb91e9d3503a760. First appeared in rclone v1.49.0. That means the handling of echo you linked to is now only needed there if we want to support rclone versions older than v1.49.0 accessing a sftp remote connecting to a rclone serve sftp instance running with newer version of rclone (e.g. latest). I'm writing a code comment regarding this - or do you think it's safe to just delete it?

> If that was activated then the md5sum test would fail and it wouldn't attempt an MD5 checksum.
> Something like this does fix the problem, perhaps in a more general way as it returns an error however md5sum is called.

I'm not sure I followed you here; seems to me that change would be good, but did you mean it should be worked on further into "a more general way" or that this *was* the more general way..?

> do we need to special case the md5sum/sha1sum detection in rclone serve sftp?

I can't think of anything, but don't know if you had anything specific in mind..

**Comment by albertony** [comment]
Testing the suggested change (with rclone serve sftp on a crypt remote)...

With an existing remote, the config will already contain:
```
md5sum_command = md5sum
sha1sum_command = sha1sum
```
(because the autodetect after v1.49 uses empty hash, and that always succeeded until the change suggested here).
And therefore rclone will still try to calculate hash, and fail (partial log output):
```
DEBUG : Running remote command
ERROR : Failed to calculate dst hash: failed to calculate md5 hash: failed to run "md5sum /path": md5 hash not supported: Process exited with status 1
ERROR : path: corrupted on transfer: md5 hash differ "cc03e747a6afcacbf5be7368acfebee5" vs ""
```
So reconfiguring the remote is required - after this fix.

When creating a new remote, first run will correctly detect that hashes are not supported, and write to config:
```
md5sum_command = none
sha1sum_command = none
```

**Comment by ncw** [comment]
I worked this up into a full patch for you to test @Forum1877621 

Note that you'll need to run the new binary as `rclone serve sftp`

And as @albertony suggested above, you'll need to remove the `md5sum_command` and `sha1sum_command` lines in the config of the sftp client otherwise it won't take effect. The sftp client doesn't need updating.

[v1.60.0-beta.6390.937f644d8.fix-6351-serve-sftp](https://beta.rclone.org/branch/fix-6351-serve-sftp/v1.60.0-beta.6390.937f644d8.fix-6351-serve-sftp/) on branch [fix-6351-serve-sftp](https://github.com/rclone/rclone/tree/fix-6351-serve-sftp) (uploaded in 15-30 mins)

@albertony do you think this fix is good enough? Anyone coming across this problem would have had this problem immediately so I hope would have told us about it.

**Comment by albertony** [comment]
> do you think this fix is good enough? Anyone coming across this problem would have had this problem immediately so I hope would have told us about it.

Yes, I agree.

**Comment by Forum1877621** [comment]
@ncw Am I missing something or is there no Windows binary?

**Comment by Forum1877621** [comment]
Looks there's no Windows binary because some test failed

https://github.com/rclone/rclone/runs/7675263694

**Comment by Forum1877621** [comment]
```
?   	github.com/rclone/rclone/cmd/mount2	[no test files]
2022/08/04 15:19:49 NOTICE: Config file "C:\\Users\\runneradmin\\AppData\\Roaming\\rclone\\rclone.conf" not found - using defaults
--- FAIL: TestRc (0.04s)
    rc_test.go:54: Mount types [cmount]
    --- FAIL: TestRc/Mount (0.04s)
        rc_test.go:93: 
            	Error Trace:	rc_test.go:93
            	Error:      	Received unexpected error:
            	            	CreateFile C:\Users\RUNNER~1\AppData\Local\Temp\TestRc4205834023\002\file.txt: The I/O operation has been aborted because of either a thread exit or an application request.
            	Test:       	TestRc/Mount
FAIL
FAIL	github.com/rclone/rclone/cmd/mountlib	0.119s
```

**Comment by ncw** [comment]
@Forum1877621 I gave that build a kick - hopefully it will produce a Windows binary in ~20mins.

**Comment by Forum1877621** [comment]
Tested and working, thanks for the fix.

**Comment by ncw** [comment]
Thanks for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.60 and v1.59.1
